lib/remote/configstageshandler.cpp

```diff
@@ -107,6 +109,13 @@ void ConfigStagesHandler::HandlePost(
 	if (!ConfigPackageUtility::ValidatePackageName(packageName))
 		return HttpUtility::SendJsonError(response, params, 400, "Invalid package name '" + packageName + "'.");
 
+	if (m_RunningPackageUpdates.exchange(true)) {
+		return HttpUtility::SendJsonError(response, params, 423,
+			"Conflicting request, there is already an ongoing package update in progress. Please try it again later.");
+	}
+
+	auto resetPackageUpdates (Shared<Defer>::Make([]() { ConfigStagesHandler::m_RunningPackageUpdates.store(false); }));
```

Comment:
I think this can be moved down even further, all the way to after https://github.com/Icinga/icinga2/blob/129b1ddc88f43b469ec9660037108ddd56a145d7/lib/remote/configstageshandler.cpp#L137-L138

Everything up to there are just fetching data from the request or checks of that data, so if any of that fails the request, there's no need to obtain the lock at all.